File: tianshu-ts/src/core/orchestrator/service.ts
import type { Agent, AgentResult, MessageBus } from "../agent/runtime.js";

export interface SubTask {
  id: string;
  worker: string;
  goal: string;
  status: string;
  result: AgentResult | null;
  error: string | null;
}

export interface Orchestration {
  task: string;
  subtasks: SubTask[];
  summary: string;
}

export type EventSinkFn = (agent: string, event: string, data: Record<string, unknown>) => Promise<void>;

export function serializeSubtask(st: SubTask): Record<string, unknown> {
  const d: Record<string, unknown> = { ...st };
  if (d.result !== null && d.result !== undefined) {
    d.result = {
      content: st.result!.content,
      tool_calls: st.result!.tool_calls,
      child_agents: st.result!.child_agents,
      error: st.result!.error,
    };
  }
  return d;
}

class Semaphore {
  private _available: number;
  private _queue: Array<() => void> = [];

  constructor(n: number) {
    this._available = Math.max(1, n);
  }

  async acquire(): Promise<void> {
    if (this._available > 0) {
      this._available--;
      return;
    }
    await new Promise<void>((resolve) => this._queue.push(resolve));
  }

  release(): void {
    const next = this._queue.shift();
    if (next) next();
    else this._available++;
  }

  async run<T>(fn: () => Promise<T>): Promise<T> {
    await this.acquire();
    try {
      return await fn();
    } finally {
      this.release();
    }
  }
}

export class Orchestrator {
  main: Agent;
  bus: MessageBus;
  parallel: boolean;
  max_workers: number | null;
  decompose_iterations: number;
  event_sink: EventSinkFn;

  constructor(
    mainAgent: Agent,
    bus: MessageBus | null = null,
    parallel = true,
    maxWorkers: number | null = null,
    decomposeIterations = 3,
    eventSink: EventSinkFn | null = null,
  ) {
    this.main = mainAgent;
    this.bus = bus ?? mainAgent.bus;
    this.parallel = parallel;
    this.max_workers = maxWorkers;
    this.decompose_iterations = decomposeIterations;
    this.event_sink = eventSink ?? (async () => {});
  }

  async run(task: string, planningPrompt = "", context = ""): Promise<Orchestration> {
    await this._emit("phase", { phase: "decompose" });
    const plan = await this.decompose(task, planningPrompt, context);
    if (!plan.subtasks.length) {
      await this._emit("phase", { phase: "direct" });
      const msg = context ? `${context}\n\n${task}`.trim() : task;
      const result = await this.main.handleMessage(msg);
      return { task, subtasks: [], summary: result.content ?? "" };
    }

    for (const st of plan.subtasks) {
      await this._emit("subtask_start", { worker: st.worker, goal: st.goal, subtask_id: st.id });
    }

    if (this.parallel) {
      const sem = new Semaphore(this.max_workers ?? plan.subtasks.length);
      await Promise.all(
        plan.subtasks.map(async (st) => {
          await sem.run(async () => {
            await this._execute(st);
            await this._emit("subtask_done", { worker: st.worker, subtask_id: st.id, status: st.status });
          });
        }),
      );
    } else {
      for (const st of plan.subtasks) {
        await this._execute(st);
        await this._emit("subtask_done", { worker: st.worker, subtask_id: st.id, status: st.status });
      }
    }

    await this._emit("phase", { phase: "aggregate" });
    plan.summary = await this.aggregate(task, plan);
    await this._emit("phase", { phase: "done" });
    return plan;
  }

  private async _emit(event: string, data: Record<string, unknown>): Promise<void> {
    try {
      await this.event_sink("orchestrator", event, data);
    } catch {
      console.error(`event_sink 异常 event=${event}`);
    }
  }

  async decompose(task: string, planningPrompt = "", context = ""): Promise<Orchestration> {
    const ctxLine = context ? `\n参考上下文:\n${context}` : "";
    const prompt =
      planningPrompt ||
      `你是任务规划者。将以下任务拆解为若干可并行或串行执行的子任务,` +
        `每个子任务需指明:worker(执行该任务的最佳 Agent 名称)、goal(清晰的目标描述)。\n` +
        `任务: ${task}\n` +
        `可用 Agent: ${this.bus.names().join(", ")}\n` +
        `请以 JSON 输出,格式: {"subtasks": [{"worker": "...", "goal": "..."}]}` +
        ctxLine;
    const result = await this.main.provider.chat([tianshuMsg(prompt)], { tools: [decomposeToolSchema()] });
    return parsePlan(task, result);
  }

  private async _execute(st: SubTask): Promise<void> {
    const worker = this.bus.agent(st.worker);
    if (!worker) {
      st.status = "failed";
      st.error = `Worker 不存在: ${st.worker}`;
      return;
    }
    try {
      st.result = await worker.handleMessage(st.goal, "orchestrator");
      st.status = st.result.error ?? "done";
      if (st.result.error) st.status = "error";
    } catch (e) {
      st.status = "failed";
      st.error = String(e);
      console.error(`子任务执行失败 worker=${st.worker} goal=${st.goal.slice(0, 100)}`);
    }
  }

  async aggregate(task: string, plan: Orchestration): Promise<string> {
    if (!plan.subtasks.length) return "";
    const parts = plan.subtasks.map(
      (st) => `[${st.worker}] ${st.goal}\n结果: ${st.result?.content ?? st.error}`,
    );
    const prompt =
      `你是汇总员。以下是主任务 '${task}' 各子任务的执行结果,请整合成一份完整、连贯的最终答复。\n\n` +
      parts.join("\n\n");
    const result = await this.main.provider.chat([tianshuMsg(prompt)]);
    return result.content ?? "";
  }
}

export function tianshuMsg(content: string): { role: "user"; content: string } {
  return { role: "user", content };
}

export function decomposeToolSchema(): Record<string, unknown> {
  return {
    type: "function",
    function: {
      name: "produce_plan",
      description: "输出子任务拆解计划",
      parameters: {
        type: "object",
        properties: {
          subtasks: {
            type: "array",
            items: {
              type: "object",
              properties: {
                worker: { type: "string" },
                goal: { type: "string" },
              },
              required: ["worker", "goal"],
            },
          },
        },
        required: ["subtasks"],
      },
    },
  };
}

export function parsePlan(task: string, result: { content: string | null; tool_calls?: Array<{ name: string; arguments: Record<string, unknown> }> | null }): Orchestration {
  const plan: Orchestration = { task, subtasks: [], summary: "" };
  const raw = result.content ?? "";
  if (result.tool_calls?.length) {
    for (const tc of result.tool_calls) {
      if (tc.name === "produce_plan") {
        const items = (tc.arguments.subtasks ?? []) as Array<{ worker?: string; goal?: string }>;
        items.forEach((item, idx) => {
          plan.subtasks.push({
            id: `st_${idx}`,
            worker: item.worker ?? "",
            goal: item.goal ?? "",
            status: "pending",
            result: null,
            error: null,
          });
        });
        return plan;
      }
    }
  }
  const m = raw.match(/\{[\s\S]*\}/);
  if (m) {
    try {
      const data = JSON.parse(m[0]) as { subtasks?: Array<{ worker?: string; goal?: string }> };
      (data.subtasks ?? []).forEach((item, idx) => {
        plan.subtasks.push({
          id: `st_${idx}`,
          worker: item.worker ?? "",
          goal: item.goal ?? "",
          status: "pending",
          result: null,
          error: null,
        });
      });
    } catch {
      // JSON 解析失败则返回空计划
    }
  }
  return plan;
}

File: tianshu/core/orchestrator/service.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Any

from tianshu.core.agent.runtime import Agent, AgentResult, MessageBus
from tianshu.core.log import get_logger

logger = get_logger("orchestrator")


@dataclass
class SubTask:
    id: str
    worker: str
    goal: str
    status: str = "pending"
    result: AgentResult | None = None
    error: str | None = None


@dataclass
class Orchestration:
    task: str
    subtasks: list[SubTask] = field(default_factory=list)
    summary: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "task": self.task,
            "subtasks": [_serialize_subtask(s) for s in self.subtasks],
            "summary": self.summary,
        }


def _serialize_subtask(st: SubTask) -> dict[str, Any]:
    d = dict(st.__dict__)
    if d.get("result") is not None:
        d["result"] = d["result"].to_dict()
    return d


class Orchestrator:
    """主 Agent: 分解任务 -> 并行或串行调度子 Agent -> 汇总结果。"""

    def __init__(
        self,
        main_agent: Agent,
        bus: MessageBus | None = None,
        parallel: bool = True,
        max_workers: int | None = None,
        decompose_iterations: int = 3,
        event_sink=None,
    ) -> None:
        self.main = main_agent
        self.bus = bus or main_agent.bus
        self.parallel = parallel
        self.max_workers = max_workers
        self.decompose_iterations = decompose_iterations
        self.event_sink = event_sink or (lambda *a, **k: None)

    async def run(self, task: str, planning_prompt: str = "", context: str = "") -> Orchestration:
        await self._emit("phase", phase="decompose")
        plan = await self.decompose(task, planning_prompt, context)
        if not plan.subtasks:
            await self._emit("phase", phase="direct")
            msg = f"{context}\n\n{task}".strip() if context else task
            result = await self.main.handle_message(msg)
            return Orchestration(task=task, summary=result.content or "")

        for st in plan.subtasks:
            await self._emit("subtask_start", worker=st.worker, goal=st.goal, subtask_id=st.id)

        if self.parallel:
            sem = asyncio.Semaphore(self.max_workers if self.max_workers else len(plan.subtasks))
            async def _one(st: SubTask) -> None:
                async with sem:
                    await self._execute(st)
                await self._emit("subtask_done", worker=st.worker, subtask_id=st.id, status=st.status)
            await asyncio.gather(*(_one(st) for st in plan.subtasks))
        else:
            for st in plan.subtasks:
                await self._execute(st)
                await self._emit("subtask_done", worker=st.worker, subtask_id=st.id, status=st.status)

        await self._emit("phase", phase="aggregate")
        summary = await self.aggregate(task, plan)
        plan.summary = summary
        await self._emit("phase", phase="done")
        return plan

    async def _emit(self, event: str, **data: Any) -> None:
        try:
            await self.event_sink("orchestrator", event, data)
        except Exception:
            logger.exception("event_sink 异常 event=%s", event)

    async def decompose(self, task: str, planning_prompt: str = "", context: str = "") -> Orchestration:
        ctx_line = f"\n参考上下文:\n{context}" if context else ""
        prompt = (
            planning_prompt
            or (
                f"你是任务规划者。将以下任务拆解为若干可并行或串行执行的子任务,"
                f"每个子任务需指明:worker(执行该任务的最佳 Agent 名称)、goal(清晰的目标描述)。\n"
                f"任务: {task}\n"
                f"可用 Agent: {', '.join(self.bus._agents.keys())}\n"
                f"请以 JSON 输出,格式: {{\"subtasks\": [{{\"worker\": \"...\", \"goal\": \"...\"}}]}}"
                f"{ctx_line}"
            )
        )
        result = await self.main.provider.chat(
            [tianshu_msg(prompt)],
            tools=[decompose_tool_schema()],
        )
        return _parse_plan(task, result)

    async def _execute(self, st: SubTask) -> None:
        worker = self.bus._agents.get(st.worker)
        if worker is None:
            st.status = "failed"
            st.error = f"Worker 不存在: {st.worker}"
            return
        try:
            st.result = await worker.handle_message(st.goal, sender="orchestrator")
            st.status = st.result.error or "done"
            if st.result.error:
                st.status = "error"
        except Exception as e:
            st.status = "failed"
            st.error = str(e)
            logger.exception("子任务执行失败 worker=%s goal=%s", st.worker, st.goal[:100])

    async def aggregate(self, task: str, plan: Orchestration) -> str:
        if not plan.subtasks:
            return ""
        parts = []
        for st in plan.subtasks:
            parts.append(f"[{st.worker}] {st.goal}\n结果: {st.result.content if st.result else st.error}")
        prompt = (
            f"你是汇总员。以下是主任务 '{task}' 各子任务的执行结果,请整合成一份完整、连贯的最终答复。\n\n"
            + "\n\n".join(parts)
        )
        result = await self.main.provider.chat([tianshu_msg(prompt)])
        return result.content or ""


def tianshu_msg(content: str):
    from tianshu.core.llm.base import LLMMessage

    return LLMMessage(role="user", content=content)


def decompose_tool_schema() -> dict[str, Any]:
    return {
        "type": "function",
        "function": {
            "name": "produce_plan",
            "description": "输出子任务拆解计划",
            "parameters": {
                "type": "object",
                "properties": {
                    "subtasks": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "worker": {"type": "string"},
                                "goal": {"type": "string"},
                            },
                            "required": ["worker", "goal"],
                        },
                    }
                },
                "required": ["subtasks"],
            },
        },
    }


def _parse_plan(task: str, result) -> Orchestration:
    plan = Orchestration(task=task)
    raw = result.content or ""
    if result.tool_calls:
        for tc in result.tool_calls:
            if tc.name == "produce_plan":
                for idx, item in enumerate(tc.arguments.get("subtasks", [])):
                    plan.subtasks.append(
                        SubTask(id=f"st_{idx}", worker=item.get("worker", ""), goal=item.get("goal", ""))
                    )
                return plan
    import json as _json
    import re as _re

    m = _re.search(r"\{.*\}", raw, _re.DOTALL)
    if m:
        try:
            data = _json.loads(m.group(0))
        except _json.JSONDecodeError:
            data = {}
        for idx, item in enumerate(data.get("subtasks", [])):
            plan.subtasks.append(
                SubTask(id=f"st_{idx}", worker=item.get("worker", ""), goal=item.get("goal", ""))
            )
    return plan